docs/source/jit_type_system_1.rst

```diff
@@ -0,0 +1,627 @@
+.. contents::
+    :local:
+    :depth: 2
+
+
+.. testsetup::
+
+    # These are hidden from the docs, but these are necessary for `doctest`
+    # since the `inspect` module doesn't play nicely with the execution
+    # environment for `doctest`
+    import torch
+
+    original_script = torch.jit.script
+    def script_wrapper(obj, *args, **kwargs):
+        obj.__module__ = 'FakeMod'
+        return original_script(obj, *args, **kwargs)
+    torch.jit.script = script_wrapper
+
+    original_trace = torch.jit.trace
+    def trace_wrapper(obj, *args, **kwargs):
+        obj.__module__ = 'FakeMod'
+        return original_trace(obj, *args, **kwargs)
+    torch.jit.trace = trace_wrapper
+
+
+.. _type_system:
+
+
+Type System (I): TorchScript Types
+==================================
+
+This section summarizes TorchScript type system. The information is based on TorchScript language reference
+and reading the implementation.
+
+Terminology
+^^^^^^^^^^^
+
+This document uses the following terminologies:
+
+::
+
+    " " represents real keywords and delimiters that are part of the syntax
+    A | B indicates either A or B
+    ( ) indicates grouping
+    [ ] indicates optional
+    A+ indicates a regular expression where term A is repeated at least once
+    A* indicates a regular expression where term A is repeated zero or more times
+
+TorchScript Types
+^^^^^^^^^^^^^^^^^
+
+The TorchScript type system consists of ``TSType`` and ``TSModuleType`` as defined below.
+
+::
+
+    TSAllType := TSType | TSModuleType
+    TSType := TSMetaType | TSPrimitiveType | TSStructuralType | TSNominalType
+
+``TSType`` represents the majority of TorchScript types, that are composable and can be used in TorchScript type annotation.
+``TSType`` can be further classified into:
+
+    * meta types, e.g., Any
+    * primitive types, e.g., int, float, str
+    * structural types, e.g., Optional[int] or List[MyClass]
+    * nominal types (Python classes), e.g., MyClass (user-defined), torch.tensor (builtin)
```

Comment:
`torch.Tensor` is not a nominal type I think.

Reply:
Hmmm, so just categorize it in builtin types?